app/app/src/main/kotlin/com/hedvig/android/app/navigation/HedvigNavHost.kt

```diff
@@ -377,7 +374,8 @@ private fun NavGraphBuilder.nestedHomeGraphs(
 }
 
 @Composable
-private fun rememberNavigator(navController: NavController): Navigator {
+private fun rememberNavigator(navController: NavController, onActivityBackPressed: () -> Unit): Navigator {
+  val updatedOnActivityBackPressed by rememberUpdatedState(onActivityBackPressed)
```

Comment:
I don't entirely understand why it's wrapped in rememberUpdatedState here:)

Reply:
I am gonna go on a trip down the memory lane and share with you this 2 year old thread https://slack-chats.kotlinlang.org/t/507148/i-have-a-question-regarding-compose-and-some-differences-bet#7ab510bc-7b86-4ce7-a8a1-75ab75c30a27
It has some extra stuff that are not exactly that, but the comment from Alex specifically is the one that explains what can go wrong if the lambda is not used this way.

So there is two right ways to write this function, and one wrong. One correct way is the one in the code right now, which is in a way "the more correct" way to do it.

The other correct way would be this
```
private fun rememberNavigator(navController: NavController, onActivityBackPressed: () -> Unit): Navigator {
  return remember(navController, onActivityBackPressed) {
    object : Navigator {
      override fun NavBackStackEntry.navigate(
        destination: Destination,
        navOptions: NavOptions?,
        navigatorExtras: androidx.navigation.Navigator.Extras?,
      ) {
        if (lifecycle.currentState == Lifecycle.State.RESUMED) {
          navigateUnsafe(destination, navOptions, navigatorExtras)
        }
      }

      override fun navigateUnsafe(
        destination: Destination,
        navOptions: NavOptions?,
        navigatorExtras: androidx.navigation.Navigator.Extras?,
      ) {
        navController.navigate(destination, navOptions, navigatorExtras)
      }

      override fun navigateUp() {
        navController.navigateUp()
      }

      override fun popBackStack() {
        if (!navController.popBackStack()) {
          onActivityBackPressed()
        }
      }
    }
  }
}
```

And the wrong way would be like this
```
private fun rememberNavigator(navController: NavController, onActivityBackPressed: () -> Unit): Navigator {
  return remember(navController) {
    object : Navigator {
      override fun NavBackStackEntry.navigate(
        destination: Destination,
        navOptions: NavOptions?,
        navigatorExtras: androidx.navigation.Navigator.Extras?,
      ) {
        if (lifecycle.currentState == Lifecycle.State.RESUMED) {
          navigateUnsafe(destination, navOptions, navigatorExtras)
        }
      }

      override fun navigateUnsafe(
        destination: Destination,
        navOptions: NavOptions?,
        navigatorExtras: androidx.navigation.Navigator.Extras?,
      ) {
        navController.navigate(destination, navOptions, navigatorExtras)
      }

      override fun navigateUp() {
        navController.navigateUp()
      }

      override fun popBackStack() {
        if (!navController.popBackStack()) {
          onActivityBackPressed()
        }
      }
    }
  }
}
```

The difference is this. If this function takes in a new instance of the lambda:

With existing solution: The latest lambda is then stored inside the `updatedOnActivityBackPressed`. So when the function in the interface is called, it will go inside `updatedOnActivityBackPressed`, get that latest version of the lambda and call it.

With the second correct solution: When a new lambda comes, the remember is re-triggered since one of the keys has changed. Then the entire `Navigator` object is destroyed and a new one is constructed, with the new instance referencing the right lambda. Then this new `Navigator` object is returned from this function, and it will trigger a recomposition basically everywhere where it is used in the hierarchy, which is more or less the entire app.
This will give the correct behavior, albeit with some extra recompositions that are not necessary which solution number 1 above solves.

With the wrong solution: The new lambda will come to this function, but since the `remember` does not have it as a key it will not be re-run, and the `Navigator` object which was created will just stay as it is. Even if you then pass in a completely new lambda that does something very different, the `Navigator` will keep calling the first instance it ever saw. Or rather, whatever the latest instance of that lambda was at the moment the last `NavController` was passed in. Since the remember **will** be re-run when that changes since it's a key in the remember.

Reply:
That's great, thanks! I think that the second right option is actually more readable - but that's probably just the question of getting used to it. And of course that one gives the extra recompositions, so. I'll read that link more thoroughly today, maybe will come up with more things to ask:)